### ISSUE TITLE
Kleki freezing during long session in Safari (iPad, Mac)

### ISSUE THREAD

**Issue by bitbof** [bitbof/klecks#6]
Multiple users have reported that Kleki (and thereby Klecks) can freeze during a longer session (30 minutes) in Safari on iPad or Mac. When freezing occurs it's impossible to save and the page needs to be reloaded.

I think it started around December 2021. I've yet to reproduce this myself.

If you've experienced this issue and have a reliable way of reproducing it please provide information. Thank you.

**Comment by bitbof** [comment]
In Safari canvases and CanvasRenderingContext2D can be null which points to a memory leak (https://stackoverflow.com/a/43482153/13941359).

Not sure if that's part of the slowdown / freeze.

**Comment by satopian** [comment]
There was information that it freezes after repeating fill 20 times on the iPad.
embedded mode Klecks. 
The canvas size is max width 800px height 800px.

**Comment by bitbof** [comment]
@satopian Thanks for the info.

I attempted to reproduce
- iPad (7th gen, 2019, iPadOS 15.4.1)
- 16 layers @ 800x800
- filled more than 20 times
The app continued to work without freezing.

It would be interesting to know their iPadOS version, from which year their model is, and if this happens consistently when filling 20 times.

**Comment by satopian** [comment]
The browser is Chrome and the iPad OS is the latest.
The iPad was released in November 2018.
The number of fills that freeze occurs is 3 to 40 times.
It seems that the issues always reappears in that user's environment.
Other operations may be causing the freeze, but it seems to avoid using it because it freezes frequently due to filling.
It hasn't been a few days since the memory leak was fixed, so I asked the user to clear the browser cache.
If I still know if the problem will recur, I will report it again.

**Comment by satopian** [comment]
I asked a user who freezes when using a fill.
The user didn't know how to clear the browser cache.
However, I was able to restart the iPad and test it.
When I tried it with the latest version of Klecks, the freeze no longer occurs.
The drawing time of the user is about 3 hours.

**Comment by bitbof** [comment]
@satopian That is good news. Let me know if it comes back.

I will have a look at Chrome on iPad. It should use the same rendering and JavaScript engine as Safari though.

**Comment by SnailPerson** [comment]
I use Kleki on both Mac and Macbook Pro with Safari.
On the Mac, the freeze occurs after around 20-30 minutes session time after zooming or undoing/redoing via buttons.
On the Macbook however, this doesn't seem to be the case as I've never encountered a freeze while using Kleki.

The Mac is a late 2017 27-inch model, with Safari ver.15.3 using MacOS 12.2.1 (Monterey)
The MacBook Pro is a late 2017 model with Safari ver.15.3 using MacOS 10.15.7 (Catalina)

So what seems to be the case is that a newer MacOS version is responsible for causing the freezes.

**Comment by SnailPerson** [comment]
Hello again, 
Ive noticed something regarding the freeze bug whilst working with Kleki today.

I used to press the undo button quite often in the past, but since I've gotten more used to using cmd+z instead, the freezes were getting less frequent.
So maybe it has something to do with the way UI buttons work? I don't remember any button other that undo/redo triggering the crash, but it might be something worth looking at.

Remember that this doesn't happen with browsers like Chrome, only on safari, so how safari handles web buttons etc might be a lead.

**Comment by SnailPerson** [comment]
Update on the situation, although seemingly less common, hitting Cmd+Z to undo a step **can still trigger a freeze.**

**Comment by SnailPerson** [comment]
What can also apparently happen, Safari instantly reloading Kleki / the site upon doing a freeze provoking action, such as undoing.
It gives the error message "this site was reloaded due to a problem", refuses to elaborate and leaves you with a white canvas.
Just lost a decent piece to this

**Comment by SnailPerson** [comment]
@bitbof no, did not have this issue for a long time now, might be for the fact that† i upgraded OS by now, so i can't speak for the older versions. But yeah, no freezing in recent versions
